lib/src/tokens.cc

```diff
@@ -0,0 +1,222 @@
+        _message(move(message)), _suggest_quotes(suggest_quotes) { }
+
+    string problem::what() const {
+        return _what;
+    }
+
+    string problem::message() const {
+        return _message;
+    }
+
+    bool problem::suggest_quotes() const {
+        return _suggest_quotes;
+    }
+
+    std::string problem::to_string() const {
+        return "'" + _what + "' (" + _message + ")";
+    }
+
+    bool problem::operator==(const token& other) const {
+        try {
+            problem const& other_problem = dynamic_cast<problem const&>(other);
+            return other_problem.what() == what() &&
+                   other_problem.message() == message() &&
+                   other_problem.suggest_quotes() == suggest_quotes();
+        } catch (bad_cast ex){
+            return false;
+        }
+    }
+
+    /** Comment token */
+    comment::comment(shared_ptr<simple_config_origin> origin, string text) :
+        token(token_type::COMMENT, move(origin)), _text(move(text)) { }
+
+    string comment::text() const {
+        return _text;
+    }
+
+    string comment::to_string() const {
+        return "'#" + _text + "' (COMMENT)";
+    }
+
+    bool comment::operator==(const token& other) const {
+        return other.get_token_type() == token_type::COMMENT && to_string() == other.to_string();
+    }
+
+    /** Double-slash comment token */
+    double_slash_comment::double_slash_comment(shared_ptr<simple_config_origin> origin, string text) :
+        comment(move(origin), move(text)) { }
+
+    string double_slash_comment::token_text() const {
+        return "//" + text();
+    }
+
+    /** Hash comment token */
+    hash_comment::hash_comment(shared_ptr<simple_config_origin> origin, string text) :
+        comment(move(origin), move(text)) { }
+
+    string hash_comment::token_text() const {
+        return "#" + text();
+    }
+
+    /** Substitution token */
+    substitution::substitution(shared_ptr<simple_config_origin> origin, bool optional,
+        vector<shared_ptr<token>> expression) : token(token_type::SUBSTITUTION, move(origin)), _optional(optional),
+        _expression(move(expression)) { }
+
+    bool substitution::optional() const {
+        return _optional;
+    }
+
+    vector<shared_ptr<token>> const& substitution::expression() const {
+        return _expression;
+    }
+
+    string substitution::token_text() const {
+        // TODO: this relies on tokenizer working to port properly
+        return to_string();
+    }
+
+    string substitution::to_string() const {
+        string result;
+        for (auto&& t : _expression) {
+            result += t->token_text();
+        }
+        return "'${" + result + "}'";
+    }
+
+    bool substitution::operator==(const token& other) const {
+        return other.get_token_type() == token_type::SUBSTITUTION &&
+               to_string() == other.to_string();
+    }
+
+    /** Singleton tokens */
+    shared_ptr<token> tokens::start_token() {
```

Comment:
All of these helpers (and any functions that directly return the result of calling them) should return const& shared_ptrs, so we avoid reference counting unless the caller needs to claim ownership of the token as well.